bson/bsoncodec/map_codec.go

```diff
@@ -252,6 +266,12 @@ func (mc *MapCodec) decodeKey(key string, keyType reflect.Type) (reflect.Value,
 		v := keyVal.Interface().(KeyUnmarshaler)
 		err = v.UnmarshalKey(key)
 		keyVal = keyVal.Elem()
+	// Try to decode encoding.TextUnmarshalers
+	case reflect.PtrTo(keyType).Implements(reflect.TypeOf((*encoding.TextUnmarshaler)(nil)).Elem()):
```

Comment:
```suggestion
	case reflect.PtrTo(keyType).Implements(textUnmarshalerType):
```

And then add this below L257:

```go
var textUnmarshalerType = reflect.TypeOf((*encoding.TextUnmarshaler)(nil)).Elem()
```

To keep the code consistent.